[#9420] Consolidate Student RESTFul API

rrtheonlyone opened the pull request:
Part of #9420

**Changes**:

| New endpoint        | Method           | Old Endpoint in v6  |
| ------------- |:-------------:| :-----:|
| /student   | DELETE | STUDENTS |
|                     | PUT      |  COURSE_STUDENT_DETAILS_EDIT |
| /students                     | DELETE      |  INSTRUCTOR_COURSE_DETAILS_DELETE_ALL_STUDENTS |
 

On top of updating the endpoints, I have:
- Refactored the edit student action (cleaned up methods)
- Changed name to UpdateStudent

darrenwee commented on src/main/java/teammates/ui/webapi/action/UpdateStudentAction.java:
```diff
@@ -0,0 +1,99 @@
+        StudentUpdateRequest updateRequest = getAndValidateRequestBody(StudentUpdateRequest.class);
+        try {
+            logic.validateSectionsAndTeams(Arrays.asList(student), student.course);
+            logic.validateSectionsAndTeams(Arrays.asList(student), student.course);
```
Do you need two of the same call?

rrtheonlyone replied:
Oh no! one of them is supposed to be `validateTeams` (have updated)

darrenwee commented on src/main/java/teammates/ui/webapi/action/UpdateStudentAction.java:
```diff
@@ -0,0 +1,99 @@
+            }
+
+            return new JsonResult(responseMessage.toString());
+        } catch (EnrollException | InvalidParametersException | EntityDoesNotExistException e) {
```
`EntityDoesNotExistException` should be caught in a separate `catch` so that you can return a `404` instead of a `400`. 

Semantically speaking, `400` indicates that the request is malformed while `404` indicates that the resource could not be found. You can have a valid request that points to a nonexistent resource.

rrtheonlyone replied:
Yeah your right! Updated

darrenwee commented on src/main/java/teammates/ui/webapi/action/UpdateStudentAction.java:
```diff
@@ -0,0 +1,99 @@
+            return new JsonResult(e.getMessage(), HttpStatus.SC_BAD_REQUEST);
+        } catch (EntityAlreadyExistsException e) {
+            return new JsonResult("Trying to update to an email that is already used",
+                    HttpStatus.SC_BAD_REQUEST);
```
(similar to the above comment) `400` indicates malformed request, but using an email that is already used alone does not constitute a malformed request. I feel that a `409 CONFLICT` would be more appropriate to return here.
https://tools.ietf.org/html/rfc7231#section-6.5.8

rrtheonlyone replied:
Done!

darrenwee commented on src/main/java/teammates/ui/webapi/action/UpdateStudentAction.java:
```diff
@@ -0,0 +1,99 @@
+                            .withComment(updateRequest.getComments())
+                            .build());
+
+            StringBuilder responseMessage = new StringBuilder("Student has been updated");
```
L67-76 can be extracted out of the `try`. It is not risky code and doesn't belong here.

rrtheonlyone replied:
Line 69 still cant! 
`logic.resetStudentGoogleId(updateRequest.getEmail(), courseId);`

This throws an exception as well :( .. kinda forces me to wrap the whole thing in a block. However, it is an improvement from https://github.com/TEAMMATES/teammates/blob/master/src/main/java/teammates/ui/webapi/action/PutCourseStudentDetailsEditAction.java.

darrenwee commented on src/main/java/teammates/ui/webapi/action/UpdateStudentAction.java:
```diff
@@ -0,0 +1,99 @@
+                boolean isSessionSummarySendEmail =
+                        getBooleanRequestParamValue(Const.ParamsNames.SESSION_SUMMARY_EMAIL_SEND_CHECK);
+
+                if (isSessionSummarySendEmail && sendEmail(courseId, updateRequest.getEmail())) {
```
I am not a big fan of using a boolean method like `sendEmail` to execute an action. Boolean methods are generally assumed to not have any side effects. You can read this for more detail: https://www.teamten.com/lawrence/programming/keep-if-clauses-side-effect-free.html

Please re-inline `sendEmail` into the `execute` method. You probably need to rearrange the message appending around after doing this.

rrtheonlyone replied:
Yeah your right, it isnt the best practice. I will update accordingly!

rrtheonlyone commented on src/test/java/teammates/test/cases/webapi/UpdateStudentActionTest.java:
```diff
@@ -63,78 +61,74 @@ public void testExecute() throws Exception {
+        assertEquals(HttpStatus.SC_OK, actionOutput.getStatusCode());
+        MessageOutput msgOutput = (MessageOutput) actionOutput.getOutput();
+        assertEquals("Student has been updated and email sent",
                 msgOutput.getMessage());
```
Changed

rrtheonlyone commented on src/test/java/teammates/test/cases/webapi/DeleteStudentActionTest.java:
```diff
@@ -39,10 +39,10 @@ protected void testExecute() {
-        assertEquals(HttpStatus.SC_OK, r.getStatusCode());
+        assertEquals(HttpStatus.SC_OK, getOutput.getStatusCode());
 
         ______TS("success: delete a student by id");
```
Yup I added a few tests and found some potential NPEs (fixed).

How should the design be? We should throw a `404 not found` if a random email/id is given right? So that the user of the API is aware that nothing happened? Or should we fail silently? (the backend logic seems to be currently doing this)

darrenwee replied:
IMO, `DELETE` endpoints _should_ fail with a 404 if a non-existent resource is specified, but opinion is divided over this because the behavior in this case is not defined by the HTTP 1.1 spec. One argument for silent failure is that deletes should be idempotent because you just want to ensure that the resource with the specified details does not exist, so repeated calls or attempting to delete something non-existent should return a 200 since the entity with those details does not exist. Tl;dr it's use-case and philosophy-dependent.

For our case, we should fail deletes silently as per our storage component policy: https://github.com/TEAMMATES/teammates/blob/master/docs/design.md#policies-1

My bad for the confusion. The negative test cases should still stay but they should always assert for 200. This will let us catch deviants trying to break the storage policy. You'll need to undo the changes you made to the action.

rrtheonlyone replied:
Okay no worries! Reverted accordingly

rrtheonlyone commented on src/main/java/teammates/ui/webapi/action/UpdateStudentAction.java:
```diff
@@ -0,0 +1,107 @@
+        } catch (EntityDoesNotExistException ednee) {
+            return new JsonResult(ednee.getMessage(), HttpStatus.SC_NOT_FOUND);
+        } catch (EntityAlreadyExistsException e) {
+            return new JsonResult("Trying to update to an email that is already used",
```
Done!

rrtheonlyone commented on src/main/java/teammates/ui/webapi/action/UpdateStudentAction.java:
```diff
@@ -0,0 +1,107 @@
+        return new JsonResult(responseMessage.toString());
+    }
+
+    private boolean sendEmail(String courseId, String studentEmail) {
```
Done!

rrtheonlyone commented on src/main/java/teammates/ui/webapi/action/UpdateStudentAction.java:
```diff
@@ -0,0 +1,107 @@
+            if (!student.email.equals(updateRequest.getEmail())) {
+                logic.resetStudentGoogleId(updateRequest.getEmail(), courseId);
+                boolean isSessionSummarySendEmail =
+                        getBooleanRequestParamValue(Const.ParamsNames.SESSION_SUMMARY_EMAIL_SEND_CHECK);
```
Alright no worries, have updated the DTO!

rrtheonlyone commented on src/main/java/teammates/ui/webapi/action/UpdateStudentAction.java:
```diff
@@ -0,0 +1,107 @@
+                    new EmailGenerator().generateFeedbackSessionSummaryOfCourse(courseId, studentEmail);
+            emailSender.sendEmail(email);
+            return true;
+        } catch (Exception e) {
```
Yes you are right, strange. This was how it was previously so I just left it in! I have removed it and updated accordingly!

rrtheonlyone commented on src/main/java/teammates/ui/webapi/action/UpdateStudentAction.java:
```diff
@@ -0,0 +1,107 @@
+                    HttpStatus.SC_CONFLICT);
+        }
+
+        StringBuilder responseMessage = new StringBuilder("Student has been updated");
```
Done!

rrtheonlyone commented:
Ready for review! Have fixed the issues as mentioned (have one query above)

darrenwee commented on src/main/java/teammates/ui/webapi/action/UpdateStudentAction.java:
```diff
@@ -0,0 +1,99 @@
+                    HttpStatus.SC_CONFLICT);
+        }
+
+        return emailSent ? new JsonResult(SUCCESSFUL_UPDATE_WITH_EMAIL) : new JsonResult(SUCCESSFUL_UPDATE);
```
I'm not a fan of the ternary. Let's do an early return instead so that it's easier to extend and modify in the future if the need ever arises.

```
if (emailSent) {
    return new JsonResult(...);
}
return new JsonResult(.....);
```

rrtheonlyone replied:
Yup updated this part!

@wkurniawan07 agreed! `emailSent == false` does mean those (2) things. However, should we update the message/status we send back to the user based on this fact?

wkurniawan07 replied:
> However, should we update the message/status we send back to the user based on this fact?

Ideally.

rrtheonlyone commented on src/main/java/teammates/ui/webapi/action/ActionFactory.java:
```diff
@@ -58,9 +58,17 @@
         map(ResourceURIs.STUDENTS, GET, GetStudentsAction.class);
-        map(ResourceURIs.STUDENTS, DELETE, DeleteStudentAction.class);
+        map(ResourceURIs.STUDENTS, DELETE,
+                DeleteInstructorCourseAllStudentsAction.class);
```
Renamed to `DeleteStudentsAction`

rrtheonlyone commented on src/main/java/teammates/ui/webapi/action/UpdateStudentAction.java:
```diff
@@ -0,0 +1,99 @@
+        if (!userInfo.isInstructor) {
+            throw new UnauthorizedAccessException("Instructor privilege is required to access this resource.");
+        }
+        String courseId = getRequestParamValue(Const.ParamsNames.COURSE_ID);
```
Yes your right! Updated

rrtheonlyone commented on src/main/java/teammates/ui/webapi/action/UpdateStudentAction.java:
```diff
@@ -0,0 +1,99 @@
+                }
+            }
+        } catch (EnrollException | InvalidParametersException e) {
+            return new JsonResult(e.getMessage(), HttpStatus.SC_BAD_REQUEST);
```
Yup it is (sry)

rrtheonlyone commented:
@wkurniawan07 okay ready for review.

- Beefed up some of the tests (Delete Students Action)
- Refactored and added new status messages for emails (not sure how to test the update success + email failure)
